crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -439,15 +443,13 @@ impl<'db> TypeInferenceBuilder<'db> {
             }
         }
 
-        if self.types.has_deferred {
-            // invariant: only annotations and base classes are deferred, and both of these only
-            // occur within a declaration (annotated assignment, function or class definition)
-            for definition in self.types.declarations.keys() {
-                if infer_definition_types(self.db, *definition).has_deferred {
-                    let deferred = infer_deferred_types(self.db, *definition);
-                    self.types.expressions.extend(&deferred.expressions);
-                    self.diagnostics.extend(&deferred.diagnostics);
-                }
+        if !self.types.deferred.is_empty() {
+            let mut deferred_inference = Vec::with_capacity(self.types.deferred.len());
+            for definition in &self.types.deferred {
+                deferred_inference.push(infer_deferred_types(self.db, *definition));
+            }
+            for deferred in deferred_inference {
+                self.extend(deferred);
             }
         }
```

Comment:
Is it possible that inferring the deferred types add new entries to `self.types.deferred`?

I would probably write this as
```suggestion
				let deferred = std::mem::take(&mut self.deferred);
				for definition in deferred {
					self.extend(infer_derferred_types(self.db, *definition);
				}
        
        assert!(self.deferred.is_empty(), "Inferring deferred types shouldn't add new types to infer"); 
```

Reply:
I agree it's important we either enforce and check that inferring deferred doesn't add more deferred, or else have explicit handling in case it does.

Reply:
Yeah, makes sense, thanks